File: reviews-service/src/main/java/com/example/reviews_service/service/ReviewService.java
package com.example.reviews_service.service;

import com.example.reviews_service.dto.CreateReviewRequest;
import com.example.reviews_service.dto.ReviewResponse;
import com.example.reviews_service.dto.UpdateReviewRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.core.Authentication;

public interface ReviewService {

    ReviewResponse createReview(CreateReviewRequest request,
                                Authentication authentication);

    ReviewResponse getReview(Long reviewId);

    Page<ReviewResponse> getProductReviews(Long productId,
                                           Pageable pageable);

    Page<ReviewResponse> getMyReviews(Authentication authentication,
                                      Pageable pageable);

    ReviewResponse updateReview(Long reviewId,
                                UpdateReviewRequest request,
                                Authentication authentication);

    void deleteReview(Long reviewId,
                      Authentication authentication);
}

File: ecommerce-frontend/src/services/reviewService.js
import {
  apiGet,
  apiPost,
  apiPatch,
  apiDelete
} from "./api";

function createReview(request) {
  return apiPost(
    "/reviews",
    request
  );
}

function getReview(reviewId) {
  return apiGet(
    `/reviews/${reviewId}`
  );
}

function getProductReviews(
  productId,
  {
    page = 0,
    size = 20,
    sort
  } = {}
) {
  const params = new URLSearchParams();

  params.set("page", page);
  params.set("size", size);

  if (sort) {
    params.set("sort", sort);
  }

  return apiGet(
    `/reviews/product/${productId}?${params.toString()}`
  );
}

function getMyReviews({
  page = 0,
  size = 20,
  sort
} = {}) {
  const params = new URLSearchParams();

  params.set("page", page);
  params.set("size", size);

  if (sort) {
    params.set("sort", sort);
  }

  return apiGet(
    `/reviews/me?${params.toString()}`
  );
}

function updateReview(
  reviewId,
  request
) {
  return apiPatch(
    `/reviews/${reviewId}`,
    request
  );
}

function deleteReview(reviewId) {
  return apiDelete(
    `/reviews/${reviewId}`
  );
}

function searchProductReviews(
  productId,
  {
    page = 0,
    size = 20,
    sort
  } = {}
) {
  const params = new URLSearchParams();

  params.set("page", page);
  params.set("size", size);

  if (sort) {
    params.set("sort", sort);
  }

  return apiGet(
    `/search/reviews/product/${productId}?${params.toString()}`
  );
}

export {
  createReview,
  getReview,
  getProductReviews,
  getMyReviews,
  updateReview,
  deleteReview,
  searchProductReviews
};